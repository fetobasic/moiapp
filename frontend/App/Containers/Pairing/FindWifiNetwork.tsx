import React, { useCallback, useRef, useState } from 'react';
import { View, ScrollView, Text, StyleSheet, Image, Linking } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import AndroidOpenSettings from 'react-native-android-open-settings';
import { NativeStackScreenProps } from '@react-navigation/native-stack';

import { HeaderSimple as Header, Button, InfoRow } from 'App/Components';
import { ApplicationStyles, Colors, Fonts, Images, isIOS, Metrics } from 'App/Themes';

import InfoIcon from 'App/Images/Icons/information.svg';
import ArrowBack from 'App/Images/Icons/arrowBack.svg';
import renderElement from 'App/Services/RenderElement';
import { useAppDispatch, usePairingNavigation } from 'App/Hooks';
import { enableForceWifiUsage } from 'App/Services/AndroidWifi';
import { showConfirm, showInfo } from 'App/Services/Alert';
import { cacheActions } from 'App/Store/Cache';
import { PairingStackParamList } from 'App/Types/NavigationStackParamList';
import DirectApi from 'App/Services/DirectApi';
import { isLegacyYeti } from 'App/Services/Yeti';
import AttentionIcon from 'App/Images/Icons/attention.svg';
import AppConfig from 'App/Config/AppConfig';

type Props = NativeStackScreenProps<PairingStackParamList, 'FindWifiNetwork'>;

const api = DirectApi.create();

function FindWifiNetwork({ route }: Props) {
  const dispatch = useAppDispatch();
  const navigation = usePairingNavigation('FindWifiNetwork');
  const isFirstRequest = useRef(true);
  const [isContinueBtnDisabled, setIsContinueBtnDisabled] = useState(true);

  const goToSettings = () => {
    if (isIOS) {
      Linking.openURL('App-Prefs:WIFI');
    } else {
      AndroidOpenSettings.wifiSettings();
    }

    setTimeout(() => {
      setIsContinueBtnDisabled(false);
    }, AppConfig.defaultDelay);
  };

  const onContinuePress = useCallback(async () => {
    enableForceWifiUsage();

    const { ok, data } = await api.getYetiInfo();

    if (data?.name && isLegacyYeti(data.name)) {
      navigation.navigate('ConnectYeti', {
        ...route.params,
        ssid: data.name,
        device: {
          thingName: data.name,
          name: data.name,
        },
        deviceType: 'YETI',
        connectionType: 'direct',
        dataTransferType: 'wifi',
      });
    } else {
      // Need to check if it's first request, because from iOS 14 we have permission request
      // and we don't want to show alert on first request
      if (!ok && isFirstRequest.current && isIOS) {
        isFirstRequest.current = false;
        return;
      }

      dispatch(cacheActions.changeAppRateInfo({ isBlockedShowAppRate: true }));

      if (data?.name && !isLegacyYeti(data.name)) {
        showInfo(
          'Your new device does not support legacy Wi-Fi.\nPlease try again using bluetooth.',
          'You have a newer device',
          () => {
            navigation.navigate('AddNewDevice');
          },
          <AttentionIcon />,
          'CONFIRM',
          'PAIR USING BLUETOOTH',
        );
      } else {
        showConfirm(
          'Connect Error',
          'We were unable to confirm connection\nto the device.',
          () => {},
          {
            cancelTitle: 'Go to settings',
            confirmTitle: 'Ok',
          },
          goToSettings,
          false,
          <AttentionIcon />,
        );
      }
    }

    isFirstRequest.current = false;
  }, [dispatch, navigation, route.params]);

  return (
    <SafeAreaView edges={['bottom']} style={ApplicationStyles.mainContainer}>
      <Header isModal title="Pair Device" leftIcon={renderElement(<ArrowBack />)} />
      <View style={ApplicationStyles.container}>
        <ScrollView showsVerticalScrollIndicator={false}>
          <InfoRow
            icon={renderElement(<InfoIcon />)}
            withBorder={false}
            title="Pair Using Device Wi-Fi"
            style={{ marginHorizontal: -Metrics.halfMargin }}
            body={renderElement(
              <>
                <Text style={styles.text}>1. Go to your phone's Wi-Fi settings.</Text>
                <Text style={styles.text}>
                  2. Select the Wi-Fi network with the name that matches the Goal Zero device you want to pair.
                </Text>
                <Text style={styles.text}>
                  <Text style={styles.textNote}>Example:</Text> Yeti9C9C1FF112
                </Text>
                <Text style={styles.text}>
                  3. Enter <Text style={{ color: Colors.green }}>"GOALZERO"</Text> when asked for a password.
                </Text>
                <Text style={styles.text}>4. Return to the Goal Zero app and continue the pairing process.</Text>
                <Text style={styles.text}>
                  <Text style={styles.textNote}>Note: </Text>After pairing, return your phone to your usual Wi-Fi
                  network.
                </Text>
              </>,
            )}
          />

          <View style={styles.imgContainer}>
            <Image
              resizeMode="contain"
              source={isIOS ? Images.selectWifiIos : Images.selectWifiAndroid}
              style={styles.img}
            />
          </View>
          <View style={styles.container}>
            <Text style={styles.textBody}>
              <Text style={[{ color: Colors.red }]}>*</Text> Newer generation Yeti models do not support Wi-Fi direct
              connect
            </Text>
          </View>
        </ScrollView>
        <View style={styles.sectionButtons}>
          <Button testID="goToSettings" style={styles.btnLeft} inverse title="Go to Settings" onPress={goToSettings} />
          <Button
            testID="onContinuePress"
            style={styles.btnRight}
            title="Continue"
            onPress={onContinuePress}
            disabled={isContinueBtnDisabled}
            inverse={isContinueBtnDisabled}
          />
        </View>
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  textBody: {
    ...Fonts.font.base.bodyOne,
    color: Colors.transparentWhite('0.87'),
  },
  textHighlight: {
    color: Colors.green,
  },
  sectionButtons: {
    flexDirection: 'row',
    paddingBottom: Metrics.smallMargin,
  },
  btnLeft: {
    flex: 1,
    marginRight: Metrics.smallMargin,
  },
  btnRight: {
    flex: 1,
    marginLeft: Metrics.smallMargin,
  },
  text: {
    ...Fonts.font.base.caption,
    marginBottom: 2,
  },
  textNote: {
    color: Colors.portWarning,
    marginTop: 2,
  },
  container: {
    alignItems: 'center',
    marginTop: Metrics.baseMargin,
  },
  imgContainer: {
    borderColor: Colors.border,
    borderWidth: 2,
    borderRadius: 8,
    overflow: 'hidden',
    height: 160,
    alignItems: 'center',
  },
  img: { top: -20 },
});

export default FindWifiNetwork;
