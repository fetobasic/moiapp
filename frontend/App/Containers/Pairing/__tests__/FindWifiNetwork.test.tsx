import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';
import { Provider } from 'react-redux';
import createMockStore from 'redux-mock-store';
import { mockedState } from 'App/Fixtures/mocks/mockedState';
import FindWifiNetwork from '../FindWifiNetwork';

const mockStore = createMockStore([]);
const store = mockStore({
  ...mockedState,
});
const props = {
  navigation: {
    navigate: jest.fn() as any,
  },
  route: { params: {} },
} as any;

describe('FindWifiNetwork', () => {
  test('renders screen', () => {
    const tree = render(
      <Provider store={store}>
        <NavigationContainer>
          <FindWifiNetwork {...props} />
        </NavigationContainer>
      </Provider>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('renders screen with buttons press', () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <NavigationContainer>
          <FindWifiNetwork {...props} />
        </NavigationContainer>
      </Provider>,
    );

    const goToSettings = getByTestId('goToSettings');
    const onContinuePress = getByTestId('onContinuePress');

    fireEvent.press(onContinuePress);
    fireEvent.press(goToSettings);
  });
});
