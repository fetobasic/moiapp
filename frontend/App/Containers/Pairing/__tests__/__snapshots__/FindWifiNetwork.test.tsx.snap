// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindWifiNetwork renders screen 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "off",
      "right": "off",
      "top": "off",
    }
  }
  style={
    {
      "backgroundColor": "rgba(27, 27, 27, 1)",
      "flex": 1,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(27, 27, 27, 1)",
          "borderBottomLeftRadius": 24,
          "borderBottomRightRadius": 24,
          "elevation": 12,
          "shadowColor": "rgba(0, 0, 0, 1)",
          "shadowOffset": {
            "height": 4,
            "width": 4,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 12,
          "zIndex": 1,
        },
        {
          "paddingTop": 0,
        },
      ]
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "height": 60,
          "justifyContent": "center",
          "marginHorizontal": 24,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        hitSlop={20}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            {
              "left": 0,
              "padding": 10,
              "position": "absolute",
            },
          ]
        }
      >
        <SvgMock />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "marginHorizontal": 48,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            {
              "color": "rgba(255, 255, 255, 1)",
              "fontFamily": "RobotoCondensed-Medium",
              "fontSize": 20,
              "fontWeight": "500",
              "lineHeight": 32,
            }
          }
        >
          Pair Device
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "rgba(27, 27, 27, 1)",
        "flex": 1,
        "paddingBottom": 0,
        "paddingHorizontal": 24,
      }
    }
  >
    <RCTScrollView
      showsVerticalScrollIndicator={false}
    >
      <View>
        <View
          style={
            [
              {
                "backgroundColor": "rgba(27, 27, 27, 1)",
                "borderRadius": 16,
                "paddingHorizontal": 12,
                "paddingVertical": 15,
              },
              undefined,
              {
                "marginBottom": 8,
                "marginHorizontal": 1,
                "paddingVertical": 16,
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "flexDirection": "row",
                },
                {
                  "marginHorizontal": -12,
                },
              ]
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "height": 32,
                  "justifyContent": "center",
                  "width": 32,
                }
              }
            >
              <SvgMock />
            </View>
            <View
              style={
                [
                  {
                    "flex": 1,
                    "paddingLeft": 12,
                    "paddingRight": 12,
                  },
                  undefined,
                  {
                    "paddingRight": 0,
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "color": "rgba(255, 255, 255, 0.87)",
                      "fontFamily": "Roboto-Medium",
                      "fontSize": 16,
                      "fontWeight": "400",
                      "lineHeight": 24,
                      "marginVertical": 4,
                    },
                    undefined,
                  ]
                }
              >
                Pair Using Device Wi-Fi
              </Text>
              <Text
                style={
                  {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 12,
                    "fontWeight": "300",
                    "lineHeight": 16,
                    "marginBottom": 2,
                  }
                }
              >
                1. Go to your phone's Wi-Fi settings.
              </Text>
              <Text
                style={
                  {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 12,
                    "fontWeight": "300",
                    "lineHeight": 16,
                    "marginBottom": 2,
                  }
                }
              >
                2. Select the Wi-Fi network with the name that matches the Goal Zero device you want to pair.
              </Text>
              <Text
                style={
                  {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 12,
                    "fontWeight": "300",
                    "lineHeight": 16,
                    "marginBottom": 2,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "rgba(255, 205, 29, 1)",
                      "marginTop": 2,
                    }
                  }
                >
                  Example:
                </Text>
                 Yeti9C9C1FF112
              </Text>
              <Text
                style={
                  {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 12,
                    "fontWeight": "300",
                    "lineHeight": 16,
                    "marginBottom": 2,
                  }
                }
              >
                3. Enter 
                <Text
                  style={
                    {
                      "color": "rgba(191, 210, 43, 1)",
                    }
                  }
                >
                  "GOALZERO"
                </Text>
                 when asked for a password.
              </Text>
              <Text
                style={
                  {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 12,
                    "fontWeight": "300",
                    "lineHeight": 16,
                    "marginBottom": 2,
                  }
                }
              >
                4. Return to the Goal Zero app and continue the pairing process.
              </Text>
              <Text
                style={
                  {
                    "color": "rgba(255, 255, 255, 0.7)",
                    "fontFamily": "Roboto-Medium",
                    "fontSize": 12,
                    "fontWeight": "300",
                    "lineHeight": 16,
                    "marginBottom": 2,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "rgba(255, 205, 29, 1)",
                      "marginTop": 2,
                    }
                  }
                >
                  Note: 
                </Text>
                After pairing, return your phone to your usual Wi-Fi network.
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "borderColor": "rgba(54, 54, 54, 1)",
              "borderRadius": 8,
              "borderWidth": 2,
              "height": 160,
              "overflow": "hidden",
            }
          }
        >
          <Image
            resizeMode="contain"
            source={
              {
                "testUri": "../../../App/Images/selectWifiIos.png",
              }
            }
            style={
              {
                "top": -20,
              }
            }
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "marginTop": 24,
            }
          }
        >
          <Text
            style={
              {
                "color": "rgba(255, 255, 255, 0.87)",
                "fontFamily": "Roboto-Medium",
                "fontSize": 14,
                "fontWeight": "400",
                "lineHeight": 20,
              }
            }
          >
            <Text
              style={
                [
                  {
                    "color": "rgba(255, 83, 29, 1)",
                  },
                ]
              }
            >
              *
            </Text>
             Newer generation Yeti models do not support Wi-Fi direct connect
          </Text>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        {
          "flexDirection": "row",
          "paddingBottom": 8,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            [
              {
                "elevation": 12,
                "shadowColor": "rgba(0, 0, 0, 1)",
                "shadowOffset": {
                  "height": 4,
                  "width": 4,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 12,
              },
              {
                "flex": 1,
                "marginRight": 8,
              },
            ],
          ]
        }
        testID="goToSettings"
      >
        <View
          style={
            {
              "borderRadius": 24,
              "overflow": "hidden",
            }
          }
        >
          <RNSVGSvgView
            bbHeight="48"
            bbWidth="100%"
            focusable={false}
            height={48}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                {
                  "flex": 0,
                  "height": 48,
                  "width": "100%",
                },
              ]
            }
            width="100%"
          >
            <RNSVGGroup
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
            >
              <RNSVGDefs>
                <RNSVGLinearGradient
                  gradient={
                    [
                      0,
                      -13224394,
                      1,
                      1776411,
                    ]
                  }
                  gradientTransform={null}
                  gradientUnits={0}
                  name="grad"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0.5"
                />
              </RNSVGDefs>
              <RNSVGRect
                fill={
                  {
                    "brushRef": "grad",
                    "type": 1,
                  }
                }
                height="48"
                propList={
                  [
                    "fill",
                  ]
                }
                width="100%"
                x="0"
                y="0"
              />
              <View
                style={
                  [
                    {
                      "borderRadius": 24,
                      "marginHorizontal": 1,
                      "marginTop": 1,
                    },
                    {
                      "backgroundColor": "rgba(27, 27, 27, 1)",
                      "height": 46,
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "paddingVertical": 12,
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "Roboto-Medium",
                            "fontSize": 14,
                            "fontWeight": "700",
                            "lineHeight": 20,
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "rgba(191, 210, 43, 1)",
                          },
                          undefined,
                        ]
                      }
                    >
                      Go to Settings
                    </Text>
                  </View>
                </View>
              </View>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": true,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "opacity": 1,
            },
            [
              {
                "elevation": 12,
                "shadowColor": "rgba(0, 0, 0, 1)",
                "shadowOffset": {
                  "height": 4,
                  "width": 4,
                },
                "shadowOpacity": 0.2,
                "shadowRadius": 12,
              },
              {
                "flex": 1,
                "marginLeft": 8,
              },
            ],
          ]
        }
        testID="onContinuePress"
      >
        <View
          style={
            {
              "borderRadius": 24,
              "overflow": "hidden",
            }
          }
        >
          <RNSVGSvgView
            bbHeight="48"
            bbWidth="100%"
            focusable={false}
            height={48}
            style={
              [
                {
                  "backgroundColor": "transparent",
                  "borderWidth": 0,
                },
                {
                  "flex": 0,
                  "height": 48,
                  "width": "100%",
                },
              ]
            }
            width="100%"
          >
            <RNSVGGroup
              fill={
                {
                  "payload": 4278190080,
                  "type": 0,
                }
              }
            >
              <RNSVGDefs>
                <RNSVGLinearGradient
                  gradient={
                    [
                      0,
                      -13224394,
                      1,
                      1776411,
                    ]
                  }
                  gradientTransform={null}
                  gradientUnits={0}
                  name="grad"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0.5"
                />
              </RNSVGDefs>
              <RNSVGRect
                fill={
                  {
                    "brushRef": "grad",
                    "type": 1,
                  }
                }
                height="48"
                propList={
                  [
                    "fill",
                  ]
                }
                width="100%"
                x="0"
                y="0"
              />
              <View
                style={
                  [
                    {
                      "borderRadius": 24,
                      "marginHorizontal": 1,
                      "marginTop": 1,
                    },
                    {
                      "backgroundColor": "rgba(27, 27, 27, 1)",
                      "height": 46,
                    },
                  ]
                }
              >
                <View
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                        "paddingVertical": 12,
                      },
                      undefined,
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "alignItems": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      style={
                        [
                          {
                            "color": "rgba(255, 255, 255, 1)",
                            "fontFamily": "Roboto-Medium",
                            "fontSize": 14,
                            "fontWeight": "700",
                            "lineHeight": 20,
                            "textTransform": "uppercase",
                          },
                          {
                            "color": "rgba(255, 255, 255, 0.38)",
                          },
                          undefined,
                        ]
                      }
                    >
                      Continue
                    </Text>
                  </View>
                </View>
              </View>
            </RNSVGGroup>
          </RNSVGSvgView>
        </View>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
